server/service/src/barcode.rs

```diff
@@ -1,25 +1,149 @@
+    fn get_barcode(
+        &self,
+        ctx: &ServiceContext,
+        id: String,
+    ) -> Result<Option<Barcode>, RepositoryError> {
+        let repository = BarcodeRepository::new(&ctx.connection);
+
+        Ok(repository
+            .query_by_filter(BarcodeFilter::new().id(EqualFilter::equal_to(&id)))?
+            .pop())
+    }
+
+    fn get_barcodes(
+        &self,
+        connection_manager: &StorageConnectionManager,
+        pagination: Option<PaginationOption>,
+        filter: Option<BarcodeFilter>,
+        sort: Option<BarcodeSort>,
+    ) -> Result<ListResult<Barcode>, ListError> {
+        let pagination = get_default_pagination(pagination, MAX_LIMIT, MIN_LIMIT)?;
+        let connection = connection_manager.connection()?;
+        let repository = BarcodeRepository::new(&connection);
+
+        Ok(ListResult {
+            rows: repository.query(pagination, filter.clone(), sort)?,
+            count: i64_to_u32(repository.count(filter)?),
+        })
+    }
+
+    fn insert_barcode(
+        &self,
+        ctx: &ServiceContext,
+        barcode: &BarcodeInput,
+    ) -> Result<Barcode, InsertBarcodeError> {
+        let result = ctx
+            .connection
+            .transaction_sync(|con| {
+                validate(con, &ctx.store_id, &barcode)?;
+                let barcode_repository = BarcodeRepository::new(con);
+                let new_barcode = BarcodeRow {
+                    id: uuid(),
+                    value: barcode.value.clone(),
+                    item_id: Some(barcode.item_id.clone()),
+                    pack_size: barcode.pack_size,
+                    manufacturer_id: None,
+                    parent_id: None,
+                };
+                barcode_repository.upsert_one(&new_barcode)?;
+                let barcode = self.get_barcode(ctx, new_barcode.id)?;
+                barcode.ok_or(InsertBarcodeError::InternalError(
+                    "Failed to read the just inserted barcode".to_string(),
+                ))
+            })
+            .map_err(|err| err.to_inner_error())?;
+        Ok(result)
+    }
+}
+pub struct BarcodeService {}
+impl BarcodeServiceTrait for BarcodeService {}
+
+fn check_barcode_does_not_exist(
+    connection: &StorageConnection,
+    barcode: &BarcodeInput,
+) -> Result<bool, RepositoryError> {
+    match barcode.pack_size {
+        Some(pack_size) => {
+            let count = BarcodeRepository::new(connection).count(Some(
+                BarcodeFilter::new()
+                    .value(EqualFilter::equal_to(&barcode.value))
+                    .pack_size(EqualFilter::equal_to_i32(pack_size)),
+            ))?;
+            Ok(count == 0)
+        }
+        None => {
+            let count = BarcodeRepository::new(connection).count(Some(
+                BarcodeFilter::new().value(EqualFilter::equal_to(&barcode.value)),
+            ))?;
+            Ok(count == 0)
+        }
+    }
+}
+
+fn check_item_exists(
```

Comment:
definitely a nit: we have three methods now, if you do global check for `fn check_item_exists`, will try to remember to bring this down a level and have one method when touching this logic in the future)